snuba/cli/query_comparer.py

```diff
@@ -139,86 +53,186 @@ def query_comparer(
-        )
-        if not mismatches_exist:
+        if compare_type == "perf" and perf_blob_exists and not override:
             continue
 
-        v1_data = query_measurement(v1_row)
-        v2_data = query_measurement(v2_row)
+        if compare_type == "data" and data_blob_exists and not override:
+            continue
 
-        mismatches.append(_create_mismatch(v1_data, v2_data))
+        result_file_format = file_manager.parse_blob_name(v1_blob)
+        v1_results = file_manager.download(v1_blob)
+        v2_results = file_manager.download(v2_blob)
+
+        perf_mismatches = []
+        data_mismatches = []
+        total_rows = 0
+        for v1_result, v2_result in itertools.zip_longest(v1_results, v2_results):
+
+            assert (
+                v1_result.query_id == v2_result.query_id
+            ), f"Invalid query_ids: {v1_result.query_id} and {v2_result.query_id} must match"
+            total_rows += 1
+            if v1_result == v2_result:
+                continue
+
+            assert isinstance(v1_result, QueryMeasurementResult)
+            assert isinstance(v2_result, QueryMeasurementResult)
+
+            perf_mismatch = create_perf_mismatch(v1_result, v2_result)
+            data_mismatch = create_data_mismatch(v1_result, v2_result)
+            if perf_mismatch:
+                perf_mismatches.append(perf_mismatch)
+            if data_mismatch:
+                data_mismatches.append(data_mismatch)
+
+        def save_comparisons(compare_type: str, mismatches: MismatchResultSet) -> None:
+            compared_file_format = FileFormat(
+                f"compared-{compare_type}",
+                result_file_format.date,
+                result_file_format.table,
+                result_file_format.hour,
+            )
+            file_manager.save(compared_file_format, mismatches, header_row=True)
+            # need the filename for the slack report
+            filename = file_manager.format_filename(compared_file_format)
+            _send_slack_report(
+                compare_type, total_rows, mismatches, result_file_format.table, filename
+            )
+
+        if compare_type == "perf":
+            save_comparisons(compare_type, perf_mismatches)
+        elif compare_type == "data":
+            save_comparisons(compare_type, data_mismatches)
+        else:
+            # no compare type == run both reports
+            save_comparisons("perf", perf_mismatches)
+            save_comparisons("data", data_mismatches)
+
+
+PERF_THRESHOLDS = {
+    "query_duration_ms": 0,
+    "read_rows": 0,
+    "read_bytes": 0,
+}
+
+DATA_THRESHOLDS = {
+    "result_rows": 0,
+    "result_bytes": 0,
+}
+
+
+def create_perf_mismatch(
+    v1_data: QueryMeasurementResult, v2_data: QueryMeasurementResult
+) -> Optional[PerfMismatchResult]:
+    d_ms_1 = v1_data.query_duration_ms
+    d_ms_2 = v2_data.query_duration_ms
+    d_ms_delta = d_ms_2 - d_ms_1
```

Comment:
The duration mismatch should be based on some percentage variance of the base query duration rather than a fixed perf threshold. Otherwise you are bound do get a lot of noise in your reports. Should we start off with a 10% variance to begin with and see what results we get back?

Reply:
added 10% but also added a check for a delta of 10ms so that really fast queries don't get included in the report if they differ slightly